ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp

```diff
@@ -72,57 +112,34 @@ namespace ngraph
-                vector<T> arg0_transpose_vec;
-                vector<T> arg1_transpose_vec;
-                vector<T> arg0_broadcast_vec;
-                vector<T> arg1_broadcast_vec;
+                std::vector<T> arg0_transpose_vec;
+                std::vector<T> arg1_transpose_vec;
+                std::vector<T> arg0_broadcast_vec;
+                std::vector<T> arg1_broadcast_vec;
 
                 // pointers to updated inputs
                 const T* arg0_update = arg0;
                 const T* arg1_update = arg1;
 
                 // vars for updated inputs shapes
-                Shape wip_arg0_shape = arg0_shape;
-                Shape wip_arg1_shape = arg1_shape;
-
-                auto get_transpose_order = [](const Shape& input_shape) {
-                    size_t rank = input_shape.size();
-                    NGRAPH_CHECK(rank > 1, "Invalid input for transpose");
-                    vector<size_t> axes_order(rank);
-                    iota(axes_order.begin(), axes_order.end(), 0);
-                    swap(axes_order[rank - 1], axes_order[rank - 2]);
-                    return AxisVector{begin(axes_order), end(axes_order)};
-                };
-
-                auto get_broadcast_axes = [](const Shape& marker_shape, const Shape& target_shape) {
-                    NGRAPH_CHECK(marker_shape.size() == target_shape.size(),
-                                 "Incompatible input shapes");
-                    AxisSet broadcast_axes;
-                    for (size_t i = 0; i < marker_shape.size(); i++)
-                    {
-                        if (marker_shape[i] == 1 && target_shape[i] != 1)
-                        {
-                            broadcast_axes.insert(i);
-                        }
-                    }
-                    return broadcast_axes;
-                };
+                Shape arg0_shape_tmp = arg0_shape;
+                Shape arg1_shape_tmp = arg1_shape;
 
                 // Perform transpose if requested
                 if (transpose_arg0 && arg0_rank > 1)
                 {
                     arg0_transpose_vec.reserve(shape_size(arg0_shape));
-                    auto axis_vector = get_transpose_order(arg0_shape);
-                    swap(wip_arg0_shape[arg0_rank - 1], wip_arg0_shape[arg0_rank - 2]);
+                    auto axis_vector = details::get_transpose_order(arg0_shape);
+                    std::swap(arg0_shape_tmp[arg0_rank - 1], arg0_shape_tmp[arg0_rank - 2]);
                     opt_kernel::reshape(reinterpret_cast<const char*>(arg0),
                                         reinterpret_cast<char*>(arg0_transpose_vec.data()),
                                         arg0_shape,
                                         axis_vector,
-                                        wip_arg0_shape,
+                                        arg0_shape_tmp,
                                         sizeof(T));
 
                     arg0_update = arg0_transpose_vec.data();
```

Comment:
This might be consider for transpose and broadcast.
```
                                // previous it was:
const T* arg0_data = arg0;      // <- arg0_update
std::vector<T> arg0_new_data;   // <- arg0_transpose_vec and arg0_broadcast_vec in one


if ( transpose /*...*/){
    std::vector<T> tmp(shape_size(arg0_shape)); //initialize (resize) vector instead reserve
    auto axis_vector = details::get_transpose_order(arg0_shape);
    std::swap(arg0_shape_tmp[arg0_rank - 1], arg0_shape_tmp[arg0_rank - 2]);
    opt_kernel::reshape(reinterpret_cast<const char*>(arg0_data),
                        reinterpret_cast<char*>(tmp.data()),
                        arg0_shape,
                        axis_vector,
                        arg0_shape_tmp,
                        sizeof(T));
    
    arg0_new_data.swap(tmp);
    arg0_data = arg0_new_data;
}
```

Reply:
Hm, so instead, we have to initialize additional `tmp` vector and `swap` data 4 times in the worst scenario.
Also broadcast data vector size is usually bigger than transposed data vector.

Reply:
I was rethinking that and I liked your idea. 
Done (5a0deb8), please take a look:)

Reply:
Nice